### PR TITLE
orderbook: get for all symbols, get by symbol, subscribe to orderbook…

### PR DESCRIPTION
**Pull request by zukiy**
… (ws)

### CODE REVIEW

File: hitbtc.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	resp, err := c.httpClient.Do(request)
+	if err != nil {
+		return response, err
+	}
+
+	response, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return response, err
+	}
+	defer resp.Body.Close()
```
Comment (avdva) [inline hitbtc.go]: Probably, we should do it on line 161, after Do, but before ReadAll?

Comment (zukiy) [reply hitbtc.go]: oops) of course you're right

File: hitbtc.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+	go func() {
+		defer conn.Close()
+		for {
+			select {
+			case <-doneChan:
+				return
+			default:
+				_, r, err := conn.ReadMessage()
+				if err != nil {
```
Comment (avdva) [inline hitbtc.go:134]: here, its better to analyze what kind of an error we got.
we the socket has been closed not by us (check for io error?), we are in infitine for loop, consuming resources.

File: hitbtc.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	if err != nil {
+		return response, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		e := Err{}
+		err = e.UnmarshalJSON(response)
+		if err != nil {
+			return response, err
```
Comment (avdva) [inline hitbtc.go]: if we got json unmarshal error, we return it, which is not very informative for the caller.
we should say, that the response code was incorrect, and (possibly) add json unmarshal error

File: types.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	TakeLiquidityRate    float64 `json:"takeLiquidityRate,string"`
+	ProvideLiquidityRate float64 `json:"provideLiquidityRate,string"`
+	FeeCurrency          string  `json:"feeCurrency"`
+}
+
+// Symbols list
+//easyjson:json
+type Symbols []Symbol
+
+type Ask struct {
```
Comment (avdva) [inline types.go]: is it used anywhere?

Comment (avdva) [reply types.go]: the same, as OBook

File: types.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+		Sequence  int64   `json:"sequence"`
+		Timestamp string  `json:"timestamp"`
+	} `json:"params"`
+}
+
+// SubscribeOrderBookResponseChan channel
+type SubscribeOrderBookResponseChan chan SubscribeOrderBookResponse
+
+//easyjson:json
+type OBook struct {
```
Comment (avdva) [inline types.go]: its better to call it `Order` )

Comment (zukiy) [reply types.go]: renamed

**Review by avdva — APPROVED** [review]

**Comment by avdva** [comment]
great!

**Pull request merged**